File: 1757-minimum-jumps-to-reach-home/minimum-jumps-to-reach-home.java
class Solution {
    public int minimumJumps(int[] forbidden, int a, int b, int x) {
        Set<Integer> vis = new HashSet<>();
        for (int ele : forbidden) {
            vis.add(ele);
        }
        return findMinJumps(a, b, x, vis);
    }

    private int findMinJumps(int a, int b, int x, Set<Integer> vis) {
        int minJumps = 0;
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[] {0, 1});
        vis.add(0);

        while (!q.isEmpty()) {
            int size = q.size();

            while (size-- > 0) {
                int[] details = q.poll();
                int pos = details[0];
                int flag = details[1];

                if (x == pos) {
                    return minJumps;
                }

                // if (vis.contains(pos)) {
                //     continue;
                // }

                // vis.add(pos);

                if (pos - b >= 0 && flag == 1 && !vis.contains(pos - b)) {
                    q.add(new int[] {pos - b, 0});
                    vis.add(pos - b);
                }

                if (pos - b <= 2000 && !vis.contains(pos + a)) {
                    q.add(new int[] {pos + a, 1});
                    vis.add(pos + a);
                }
            }

            minJumps++;
        }

        return -1;
    }
}

File: 1757-minimum-jumps-to-reach-home/minimum-jumps-to-reach-home.cpp
class Solution {
public:
    int bfs(std::unordered_set<int>& vis, int a, int b, int x) {
        std::queue<std::pair<int, int>> q;
        q.push({0, 1});

        int ans = 0;
        while (!q.empty()) {
            int n = q.size();

            for (int i = 0; i < n; i++) {
                int curr = q.front().first;
                int flag = q.front().second;
                q.pop();
                if (curr == x) return ans;
                if (vis.count(curr)) continue;

                vis.insert(curr);
                if (flag == 1 && (curr - b >= 0)) q.push({curr - b, 0});

                if ((curr - b) <= 2000) q.push({curr + a, 1});
            }
            ans++;
        }
        return -1;
    }

    int minimumJumps(std::vector<int>& forbidden, int a, int b, int x) {
        std::unordered_set<int> vis;
        for (auto& z : forbidden) vis.insert(z);

        return bfs(vis, a, b, x);
    }
};
